fdb-relational-api/src/main/java/com/apple/foundationdb/relational/api/options/CollectionContract.java

```diff
@@ -24,21 +24,26 @@
 import com.apple.foundationdb.relational.api.exceptions.ErrorCode;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import java.sql.SQLException;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 /**
  * Option contract that ensures that all elements of a collection match some contract.
  * The contract is only valid if the elements of the given option are (1) some kind
  * of collection type and (2) every element in the collection satisfies the "element
  * contract" over which the contract is defined. For example, the child contract
  * may be that each element is of a given type or within a particular range.
+ * @param <T> the type parameter of the collection
  */
-public class CollectionContract implements OptionContract {
+public class CollectionContract<T> implements OptionContract, OptionContractWithConversion<Collection<T>> {
     @Nonnull
-    private final OptionContract elementContract;
+    private final TypeContract<T> elementContract;
```

Comment:
In theory, this should be an `OptionContractWithConversion<T>`, I think, which would let us do things like nesting. Though right now the serialization format of collection contracts don't support that.

This is probably fine for now, though